Streamline Parsing Logic

pi1024e opened the pull request:
<!-- What's in this pull request? -->
The ParseType code has a lot of jumps and incongruity in respect to the code's behavior despite there being functions that have identical behavior. For readability, and to highlight this, I refactored this file to better reflect that.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

lattner committed:
[Concurrency] Implement parsing and semantic analysis of await operator

Similar to `try`, await expressions have no specific semantics of their
own except to indicate that the subexpression contains calls to `async`
functions, which are suspension points. In this design, there can be
multiple such calls within the subexpression of a given `await`.

Note that we currently use the keyword `__await` because `await` in
this position introduces grammatical ambiguities. We'll wait until
later to sort out the specific grammar we want and evaluate
source-compatibility tradeoffs. It's possible that this kind of prefix
operator isn't what we want anyway.

Commit:
Update SwiftSyntaxParser.h

jckarter committed:
Add PruneVTables to the performance optimizer passes.

jckarter committed:
PruneVTables: Keep SILModule's vtable lookup table in sync with updates.

And add a verifier check to ensure the cache remains synced.

jckarter committed:
PruneVTables: Add DEBUG logging of the pass decisions

rintaro commented on lib/Parse/ParseType.cpp:
```diff
@@ -662,9 +663,7 @@ ParserStatus Parser::parseGenericArguments(SmallVectorImpl<TypeRepr *> &Args,
       // Parse the comma, if the list continues.
-      if (!consumeIf(tok::comma))
-        break;
-    }
+    } while (consumeIf(tok::comma));
```
This changes the behavior. `ElementContext` is a RAII object and it needs `consumeIf(tok::comma)` within the lifetime of it. Apparently, we don't have any test cases for it. Could you add something like:
```
typealias L = A<B, C>
```
to [test/Syntax/round_trip_parse_gen.swift](https://github.com/apple/swift/blob/master/test/Syntax/round_trip_parse_gen.swift#L146)?

pi1024e replied:
Sure!

rintaro reviewed (changes requested):
Nice cleanup! Could you address the following issue?

pi1024e commented:
@rintaro I did!

rintaro committed:
[AST] Add a pointer to parent 'TapExpr' expression in 'VarDecl'

Preparation for on-demand 'VarDecl' type checking in interpolated string
literals.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -650,7 +649,7 @@ ParserStatus Parser::parseGenericArguments(SmallVectorImpl<TypeRepr *> &Args,
         SyntaxKind::GenericArgumentList);
 
-    while (true) {
+    for (;;) {
```
Let's leave these alone; one is not better than the other, and `while (true)` is used commonly in this codebase, so it's best not to cause churn.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -720,7 +720,7 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
   SmallVector<ComponentIdentTypeRepr *, 4> ComponentsR;
   SourceLoc EndLoc;
-  while (true) {
+  for (;;) {
```
Ditto here.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -756,28 +756,26 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
-        // Consume the period.
+      if (peekToken().isContextualKeyword("Type") ||
+          peekToken().isContextualKeyword("Protocol"))
+        break;
```
👍  
This does read better.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -756,28 +756,26 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
+          break;
+      }
+      // Consume the period.
         consumeToken();
```
You'll need to de-indent, since you got rid of one level of nesting.

```suggestion
      consumeToken();
```

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -756,28 +756,26 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
+      }
+      // Consume the period.
         consumeToken();
         continue;
```
```suggestion
      continue;
```

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -756,28 +756,26 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
-        Status.setHasCodeCompletion();
-      break;
     }
+    if (Tok.is(tok::code_complete) && !Tok.isAtStartOfLine())
```
This doesn't seem to preserve the same behavior (or if somehow it does, it is not obvious in this snippet and would require non-local reasoning to prove). We want to `break` _only if_ `Tok.is(tok::code_complete)`. Please restore the `else if`.

pi1024e replied:
The continue statement itself makes the else if redundant.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1488,22 +1486,21 @@ bool Parser::canParseAsGenericArgumentList() {
-  } else {
-    consumeStartingGreater();
-    return true;
+  if (startsWithLess(Tok)) {
```
As we've discussed in the past, it is intentional to prefer early return. I think this should restore the original `if (!startsWithLess(Tok)) return false;`.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1488,22 +1486,21 @@ bool Parser::canParseAsGenericArgumentList() {
+      // Parse the comma, if the list continues.
+    } while (consumeIf(tok::comma));
+
+    if (startsWithGreater(Tok)) {
```
For the same reason, please restore the original `if (!startsWithGreater(Tok)) return false`.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1551,7 +1548,7 @@ bool Parser::canParseType() {
 
   // '.Type', '.Protocol', '?', and '!' still leave us with type-simple.
-  while (true) {
+  for (;;) {
```
See above.

```suggestion
  while (true) {
```

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1580,9 +1577,7 @@ bool Parser::canParseType() {
-    if (!canParseType())
-      return false;
-    return true;
+    return canParseType();
```
👍  
Good point.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1592,16 +1587,14 @@ bool Parser::canParseTypeIdentifierOrTypeComposition() {
     return canParseOldStyleProtocolComposition();
   
-  while (true) {
+  for (;;) {
```
See above.

```suggestion
  while (true) {
```

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1592,16 +1587,14 @@ bool Parser::canParseTypeIdentifierOrTypeComposition() {
-      consumeToken();
-      continue;
-    } else {
+    if (!Tok.isContextualPunctuator("&")) {
```
Nit: doesn't need braces, if we're going to be consistent here.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
 
 bool Parser::canParseTypeIdentifier() {
-  while (true) {
+  for (;;) {
```
See above.

```suggestion
  while (true) {
```

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
-        !peekToken().isContextualKeyword("Protocol")) {
-      consumeToken();
-    } else {
+    if (!(Tok.is(tok::period) || Tok.is(tok::period_prefix)) ||
```
Two reasons to restore the original version here: (1) this version no longer matches the logic that's described in the comment above it, but rather inverts it; (2) the same expression found in the original version is used elsewhere, so it is easy to see at a glance that it's testing for the same conditions as in those other places, whereas it is much more difficult for a reader to appreciate that the inverted logic here has the same intended effect. In fact, I cannot at a glance be sure that the logic has been inverted correctly.

For both these reasons, I'd strongly suggest restoring the original version.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
   // Qualified name base types must be followed by a period.
   // If the next token starts with a period, return true.
-  return startsWithSymbol(Tok, '.');
+  return canParseSimpleTypeIdentifier() && startsWithSymbol(Tok, '.');
```
Here too, the comments no longer describe the code. Please restore the original code or update the comments.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
-  do {
-    if (!canParseTypeIdentifier()) {
-      return false;
+  if (startsWithLess(Tok)) {
```
See above: please restore the original early return.

xwu commented on lib/Parse/ParseType.cpp:
```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
+    } while (consumeIf(tok::comma));
+
+    // Check for the terminating '>'.
+    if (startsWithGreater(Tok)) {
```
See above; please restore the original.

xwu commented on test/Syntax/round_trip_parse_gen.swift:
```diff
@@ -143,6 +143,7 @@ typealias H = () rethrows -> ()
 typealias I = (A & B<C>) -> C & D
 typealias J = inout @autoclosure () -> Int
 typealias K = (@invalidAttr Int, inout Int, __shared Int, __owned Int) -> ()
+typealias L = A<B, C>
```
What is the meaning of this added line of code?

rintaro replied:
@xwu I asked to add it because we didn't have test cases for TypeSyntax with multiple generic arguments.

@pi1024e You probably need to update `test/Syntax/Outputs/round_trip_parse_gen.swift.withkind`. Please test it locally.

xwu commented:
@pi1024e While consistent formatting is very good, please only format your own changes. It is difficult to review those changes when there are now hundreds of whitespace differences in unrelated code.

xymus committed:
[Sema] Only warn when exporting implementation-only types in SPI

Downgrade to a warning the diagnostics on the use of an
implemenetation-only type in SPI to allow support for
-experimental-spi-imports.

jckarter committed:
IRGen: Elide nonoverridden entries from public resilient vtables.

A formally virtual method still needs to provide the ABI of an overridable
method, including a dispatch thunk, method descriptor, and support in the
method lookup function for the class to handle `super.` calls from clients.

DougGregor committed:
Forward matching of trailing closure arguments.

Introsuce a new "forward" algorithm for trailing closures where
the unlabeled trailing closure argument matches the next parameter in
the parameter list that can accept an unlabeled trailing closure.

The "can accept an unlabeled trailing closure" criteria looks at the
parameter itself. The parameter accepts an unlabeled trailing closure
if all of the following are true:

* The parameter is not 'inout'
* The adjusted type of the parameter (defined below) is a function type

The adjusted type of the parameter is the parameter's type as
declared, after performing two adjustments:

* If the parameter is an @autoclosure, use the result type of the
parameter's declared (function) type, before performing the second
adjustment.
* Remove all outer "optional" types.

For example, the following function illustrates both adjustments to
determine that the parameter "body" accepts an unlabeled trailing
closure:

    func doSomething(body: @autoclosure () -> (((Int) -> String)?))

This is a source-breaking change. However, there is a "fuzzy" matching
rule that that addresses the source break we've observed in practice,
where a defaulted closure parameter precedes a non-defaulted closure
parameter:

    func doSomethingElse(
       onError: ((Error) -> Void)? = nil,
       onCompletion: (Int) -> Void
    ) { }

    doSomethingElse { x in
      print(x)
    }

With the existing "backward" scan rule, the trailing closure matches
onCompletion, and onError is given the default of "nil". With the
forward scanning rule, the trailing closure matches onError, and there
is no "onCompletion" argument, so the call fails.

The fuzzy matching rule proceeds as follows:
* if the call has a single, unlabeled trailing closure argument, and
* the parameter that would match the unlabeled trailing closure
argument has a default, and
* there are parameters *after* that parameter that require an argument
(i.e., they are not variadic and do not have a default argument)

then the forward scan skips this parameter and considers the next
parameter that could accept the unlabeled trailing closure.

Note that APIs like doSomethingElse(onError:onCompletion:) above
should probably be reworked to put the defaulted parameters at the
end, which works better with the forward scan and with multiple
trailing closures:

    func doSomethingElseBetter(
       onCompletion: (Int) -> Void,
       onError: ((Error) -> Void)? = nil
    ) { }

    doSomethingElseBetter { x in
      print(x)
    }

    doSomethingElseBetter { x in
      print(x)
    } onError: { error in
      throw error
    }

rintaro committed:
[AST] Add a enum for type checking state of ClosureExpr's body

rintaro committed:
[TypeChecker] Invalidate decls in foreach statement

If it's failed to type check the patterns, the sequence expression,
etc., invalidate the pattern and 'VarDecl's in the pattern.

rintaro committed:
[TypeChecker] Factor out StmtChecker::typeCheckASTNode()

For type checking single element in brace statements.

DougGregor committed:
Make the argument-label diagnostics slightly more robust.

If we're matching the trailing closure at the end but haven't seen any
mismatches yet, consider the trailing closure to be a mismatch.
This function probably needs to be rewritten in terms of
matchCallArguments(), because it is performing an incorrect
approximation of label matching that doesn't work for multiple
trailing closures.

rintaro committed:
[TypeChecker] Consolidate StmtChecker constructors

rintaro committed:
[CodeCompletion] Only type check related statements in function body

Introduce 'TypeCheckSingleASTNode' mode that only type checks single body
element and dependent necessities (i.e. referencing ValueDecls and their
dependencies).

Renamed swift::typeCheckAbstractFunctionBodyAtLoc() to
swift::typeCheckASTNodeAtLoc(DeclContext *, SourceLoc). That type checks
innermost 'ASTNode' at the location. Also, 'TypeCheckSingleASTNode' mode
skips type checking any "body" of the node (i.e. BraceStmt elements for
function body, if statement body, closure body, etc.)

Added on-demand type checking using it:
 - VarDecl in TapExpr
 - ParamDecl in ClosureExpr
 - Return type of ClosureExpr
 - Binding value in control statements
   (e.g. ForEachStmt, SwitchStmt, DoCatchStmt, etc.)

rdar://problem/63932852

rintaro committed:
[Sema] Remove StmtChecker.TargetTypeCheckLoc

The location itself is not used anymore. Use 'TypeCheckSingleASTNode'
instead.

rintaro committed:
[CodeCompletion] Propagate "skipping body" flag per request basis

Instead of using a global type checker flag.

rintaro committed:
[Sema] Remove on-demand ParamDecl checking for closures

Because they should have been type checked before the request.

rintaro committed:
[AST] Rename TypeCheckedAtOnce to TypeCheckedWithSignature

To clarify the meaning

rintaro committed:
[Sema] Stop on-demand type checking for 'VarDecl' in 'TapExpr'

rintaro committed:
[Sema] Remove LeaveFunctionBodyUnchecked flag from DeclChecker

At least for now, ide::typeCheckContextAt() doesn't call
'typeCheckASTNodeAtLoc()' with function decl signature position.
Also, 'getIntefaceType()' on function reference doesn't call decl
checker. So we don't need to propagate it.

rintaro committed:
[Sema] Remove on-demand type checking for return type of ClosureExpr

Closure singatures should be type checked at that point anyway.

DougGregor committed:
Extend the "fuzzy" forward scan matching to support multiple trailing closures

The "fuzzy" forward scan matching algorithm was only applied when there
was a single, unlabeled trailing closure, but was disabled in the
presence of multiple trailing closures. Extend the "fuzzy" match to
account for multiple trailing closures, by restricting the search for
"a later parameter that needs an argument" to stop when we find a
parameter that matches the first (labeled) trailing closure.

DougGregor committed:
[Trailing closures] Allow the unlabeled closure for variadic closure parameters.

Once the first argument for a variadic function-typed parameter has been
matched, allow an unlabeled trailing closure to match, rather than
banning all uses of the unlabeled trailing closure with variadic
parameters.

DougGregor committed:
Lacking parameter declarations, and parameter accepts trailing closures

When we are performing a call through a value of function type, rather
than calling a particular declaration that has parameter declarations,
allow any parameter to accept trailing closures.

DougGregor committed:
[Type checker] Improve diagnostics for trailing closures.

The change to the forward-scanning rule regressed some diagnostics,
because we no longer generated the special "trailing closure mismatch"
diagnostic. Reinstate the special-case "trailing closure mismatch"
diagnostic, but this time do so as part of the normal argument
mismatch diagnostics so it is based on type information.

While here, clean up the handling of missing-argument diagnostics to
deal with (multiple) trailing closures properly, so that we can (e.g)
suggest adding a new labeled trailing closure at the end, rather than
producing nonsensical Fix-Its.

And, note that SR-12291 is broken (again) by the forward-scan matching
rules.

DougGregor committed:
[Diagnostics] Remove the now-unused AllowInvalidUseOfTrailingClosure

Diagnosis for invalid uses of trailing closures has been folded in
with argument-matching diagnostics, so remove all of the machinery
around the syntactic "mismatched trailing closure" logic.

DougGregor committed:
[Trailing closures] Enable "fuzzy" heuristic by default everywhere.

SE-0286 states that the "fuzz" heuristic is part of the new language
behavior, so don't automatically disable it in Swift 6+ mode.

DougGregor committed:
[Trailing closures] Diagnose the change in behavior with SE-0268.

SE-0248 changes the backward-scan matching behavior for the unlabeled
trailing closure into a forward scan. In circumstances where this
could silently change the meaning of a call to a particular
function, i.e., when there are two defaulted closure parameters such
that a given closure to match either one of them, produce an warning
that describes the change in behavior. For example:

    t4.swift:2:24: warning: since Swift 5.3, unlabeled trailing
closure argument matches parameter 'x' rather than parameter 'z'
    trailingClosureSingle2 { $0 }
                           ^
    t4.swift:2:24: note: label the argument with 'z' to retain the
pre-Swift 5.3 behavior
    trailingClosureSingle2 { $0 }
                           ^
                          (z:    )
    t4.swift:2:24: note: label the argument with 'x' to silence this
warning for Swift 5.3 and newer
    trailingClosureSingle2 { $0 }
                           ^
                          (x:    )
    t4.swift:1:6: note: 'trailingClosureSingle2(x:y:z:)' contains
defaulted closure parameters 'x' and 'z'
    func trailingClosureSingle2(x: (Int) -> Int = { $0 } , y: (Int) ->
Int = { $0 }, z: (Int) -> Int = { $0 }) {}
         ^                      ~

This explains the (rare) case where SE-0286 silently changes the
meaning of a program, offering Fix-Its to either restore the
pre-SE-0286 behavior or silence the warning, as appropriate.

DougGregor committed:
Fix a NULL pointer dereference and update test cases.

DougGregor committed:
[Cleanup] Use swift::getParameterList() rather than hard-coding it.

xymus committed:
[ModuleInterface] Print some implementation-only imports in the private interface

Print implementation-only imports in the private textual interface
only if also importing SPI. This allows to export types from
implementation-only imports in SPI and brings the private textual
interfaces in line with the binary interfaces.

This is a temporary solution as we need to better design the language
feature around this.

This feature requires passing -experimental-spi-imports to the frontend
that generates the private swiftinterface file.

artemcm committed:
Factor out Explicit Module Map parsing into a standalone utility.

artemcm committed:
Add option to accept external module dependency map as input
With: `-external-dependency-module-map-file`

artemcm committed:
Add new dependency kind (swiftExternal) to model external dependencies.

artemcm committed:
Add an external module dependency test

artemcm committed:
Refactor ModuleDependenciesStorageBase to self-identify with ModuleDependenciesKind

artemcm committed:
Keep the memory allocator local to module loaders, instead of module map parser.

artemcm committed:
Rename external dependencies to placeholder dependencies.

artemcm committed:
Add in-source rationale for why Placeholder dependencies are needed.

DougGregor committed:
[Trailing closures] Attempt both forward and backward scans.

To better preserve source compatibility, teach the constraint
solver to try both the new forward scanning rule as well as the
backward scanning rule when matching a single, unlabeled trailing
closure. In the extreme case, where the unlabeled trailing closure
matches different parameters with the different rules, and yet both
produce a potential match, introduce a disjunction to explore both
possibilities.

Prefer solutions that involve forward scans to those that involve
backward scans, so we only use the backward scan as a fallback.

DougGregor committed:
[Trailing closures] Warn about use of deprecated "backward" scan.

Whenever we form a call that relies on the deprecated "backward" scan,
produce a warning to note the deprecation along with a Fix-It to label
the parameter appropriately (and suppress the warning). For example:

    warning: backward matching of the unlabeled trailing closure is
        deprecated; label the argument with 'g' to suppress this warning
      trailingClosureEitherDirection { $0 * $1 }
                                     ^
                                    (g:         )

DougGregor committed:
[Trailing closures] Reinstate the "skip defaulted argument" heuristic.

My experiment to improve source compatibility by also performing a
backward scan removed the SE-0286 heuristic that skipped binding
the unlabeled trailing closure to a defaulted parameter when that
would fail. Reinstate that heuristic, which makes more existing code
work with the forward-scan behavior.

This makes my source-compatibility improvements a quality-of-implementation

slavapestov committed:
AST: GenericParamList::clone() doesn't need to preserve source locations

slavapestov committed:
AST: Add GenericParamList::setDeclContext()

slavapestov committed:
AST: Add a way to distinguish a parsed generic parameter list from a synthesized one

Add a new GenericContext::getParsedGenericParams(). This produces
the same value as GenericContext::getGenericParams() if the generic
parameter list was written in source. For extensions and protocols,
this returns nullptr without synthesizing anything.

DougGregor committed:
[Trailing closures] Bias toward the backward scan for ambiguities.

This approach, suggested by Xiaodi Wu, provides better source
compatibility for existing Swift code, by breaking ties in favor of the
existing Swift semantics. Each time the backward-scan rule is needed
(and differs from the forward-scan result), we will produce a warning
+ Fix-It to prepare for Swift 6 where the backward rule can be
removed.

DougGregor committed:
[Trailing closures] Remove dynamically-dead warning about behavior change.

With the constraint solver preferring backward scanning to forward
scanning, there is no need to point out the ambiguity: we will
always, consistently warn about backward scanning when it produced a
result that was different from the forward scan.

slavapestov committed:
AST: Move GenericParamListRequest::evaluate() to NameLookup.cpp

jckarter committed:
IRGen: Keep vtables entries inherited from fragile base classes.

If a subclass is resilient, but has a fragile base class, we need to preserve the ABI-exposed vtable entries
inherited from the base class.

slavapestov committed:
AST: Remove redundant cloneGenericParams() helper function

jckarter committed:
IRGen: Method implementations for elided vtable entries are direct and don't need ptrauth

hamishknight committed:
[TBDGen] Factor out some visiting logic

Factor out the visiting logic for `FileUnit` and
`TBDGenDescriptor` onto the `TBDGenVisitor`.

hamishknight committed:
[TBDGen] Split off symbol collection from TBD file generation

Refactor `TBDGenVisitor` to accept a callback for
when it discovers a symbol, and split off public
symbol gathering into `PublicSymbolsRequest` such
that we don't need to unnecessarily also build a
TBD file which we immediately throw away.

hamishknight committed:
[Frontend] Fix error detection

Seperating out TBD file generation uncovered a bug
in the frontend where we could fail to notice that
we emitted an error during the end-of-pipeline
actions and returned a zero exit code. Tweak the
logic to detect this case.

gottesmm committed:
[sil] Add an implementation of isIndirectResultOperand() onto ApplySite that returns false for partial_apply.

Sometimes when working with the operands of an ApplySite, one needs to bail
early if one has a full apply site and the operand is an indirect result
argument. Sadly, there isn't any such API on ApplySite (it is only on
FullApplySite), even though technically a partial_apply can be viewed as just
not having any indirect result operands! That is what this patch implements.

To implement this I added a simple useful method called:

```
Optional<FullApplySite> ApplySite::asFullApplySite() const
```

Seems like a useful thing to have for these sorts of cases.

slavapestov committed:
AST: Don't attach trailing where clause requirements to the GenericParamList

Previously we had two representations for the 'where' clause of a
parsed declaration; if the declaration had generic parameters of
its own, we would store them in the GenericParamList, otherwise
we would store them separately in a TrailingWhereClause instance.

Since the latter is more general and also used for protocols and
extensions, let's just use it for everything and simplify
GenericParamList in the process.

shahmishal committed:
[XFAIL] Sema/SwiftUI/rdar57201781.swift and Constraints/trailing_closures_objc.swift to support Xcode 12 beta 3

varungandhi-apple committed:
[docs] Add comments clarifying roles of Type.h and Types.h.

varungandhi-apple committed:
[Gardening] Update clang-function-type-serialization comments.

We have landed support for serialization Clang function types, but there is
still work to be done here.

Moreover, we should use a consistent style with `[(FIXME|TODO|NOTE): label]`,
instead of using different styles throughout the code.

varungandhi-apple committed:
[Gardening] Use consistent style for referencing notes.

varungandhi-apple committed:
[DebugInfo] Fix test REQUIRES to run on Windows.

gottesmm committed:
[opt-remark] Add support for simple object projections.

This ensures that we are able to properly look through struct_extract,
tuple_extract in cases where we have an aggregate with multiple non-trivial
subtypes (implying it is not-rc identical with those sub-types).

The end result is that we now emit good diagnostics for things like this:

```
func returnStructWithOwnerOwner(x: StructWithOwner) -> Klass {
    return x.owner // expected-remark {{retain}}
                   // expected-note @-7:33 {{of 'x.owner'}}
}
```

eeckstein committed:
stdlib: changes for the StringOptimization

To be able to constant fold string interpolation, the right semantic attributes must be in place.
Also, the interpolation's write function must be inlinable.
For the _typeName constant folding, a semantic attribute is required.

eeckstein committed:
SIL optimizer: Add a new string optimization.

Optimizes String operations with constant operands.

Specifically:
  * Replaces x.append(y) with x = y if x is empty.
  * Removes x.append("")
  * Replaces x.append(y) with x = x + y if x and y are constant strings.
  * Replaces _typeName(T.self) with a constant string if T is statically known.

With this optimization it's possible to constant fold string interpolations, like "the \(Int.self) type" -> "the Int type"

This new pass runs on high-level SIL, where semantic calls are still in place.

rdar://problem/65642843

Commit:
temp

compnerd committed:
test: decode the output of `lldb -P` as UTF-8

We currently use the byte string as the PYTHONPATH with Python 3, which
is odd.

compnerd committed:
build: remove tensorflow-swift-apis from the master branch

Remove the ability to build this dependency.

CodaFi commented on CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 cmake_minimum_required(VERSION 3.12.4)
 
+SET(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -mtune=native -march=native -flto")
```
Let's not add optimization settings to this patch.

hyp committed:
[build-script] copy over the simulator libclang_rt.*.a libraries to allow clang to link with them

Clang's driver started linking with libclang_rt.<os>sim.a when building for simulator.

Swift's build need to adapt to this clang change, by ensuring that the libclang_rt.<os>sim.a libraries are created during the build

eeckstein committed:
ASTMangler: add a little utility function to create a mangled name for _typeName constant propagation.

eeckstein committed:
Demangler: add an option to print a demangled type name exactly as the runtime function _typeName would do.

slavapestov committed:
Sema: Fix some crashes with invalid nesting of extensions and protocols

Make sure we consistently use getParentForLookup() and not getParent()
when looking at generic DeclContexts. This is because an extension or
protocol that is nested inside of another generic context must never
inherit generic parameters from the parent context.

We already had this invariant enforced in some places, but now that we
do it more consistently we can fix more crashes of this kind.

Fixes <rdar://problem/58813746>, <https://bugs.swift.org/browse/SR-13004>.

aschwaighofer committed:
Convert the EagerSpecializer pass to a function pass

Should be NFC.

varungandhi-apple committed:
[NFC] Rename ExtInfo::Uncommon to ExtInfo::ClangTypeInfo.

The previous name was poorly chosen (by me). Time to fix that.

slavapestov committed:
Sema: Refactor exportability checking to not use the GenericRequirementsCheckListener

nkcsgexi committed:
Front-end: sink SDK version implementation to lib/Basic. NFC

augusto2112 committed:
Enable validation reflection tests passing on Linux (#33134)

slavapestov committed:
AST: Explicitly encode that only protocol and extension members are visible in a 'where' clause

Today, ASTScope only creates NominalTypeWhereScopes for the 'where'
clause of an extension or a protocol. All other generic declarations
model the 'where' clause as part of the declaration's scope itself
(but it's not part of the body scope).

Lookups into the NominalTypeWhereScope would look inside the type,
because this is how we want protocol and protocol extension scopes
to work -- you should be able to refer to associated types without
a 'Self' prefix.

Let's add a conditional check for protocols and extensions to the
implementation of NominalTypeWhereScope, so that we can use this
scope for all other 'where' clauses and keep the old behavior.

Note that today, even _concrete_ extensions have this behavior,
which is perhaps slightly unexpected:

class C<T> {
  typealias U = Int
}

extension C where T == U {}

It would be nice to tighten up the rule here but there's already a
test in the suite that depends on it.

natecook1000 committed:
Only include _getMetadataSection functions in stdlib builds w/ assertions

theblixguy committed:
[Parse] [Sema] Update confusables diagnostic to mention the character names as well (#33105)

* [Parser] Update 'Confusables.def' file to include confusable and base character names

* [Parser] Add a new utility method to return the names of the confusable and base characters for a given confusable codepoint

* [Parser] Update diagnostic for confusable character during lexing to mention confusable and base character names

* [Sema] If there is just a single confusable character, emit a tailored diagnostic that also mentions the character names

* [Diagnostics] Add new diagnostic messages to the localization file

* [Test] Update confusables test

* [Utils] Update unicode confusables txt file and update script to regenerate confusables def file

* [Parse] Regenerate 'Confusables.def' using updated script

* [Utils] Adjust generate_confusables script based on review feedback

Fix a mistake with name mapping. Updated header comment. Fix a couple of linting issues.

* [Parse] Regenerate 'Confusables.def' file once again after script changes

* [Parse] Add the newline after end of 'getConfusableAndBaseCodepointNames' method

* [Test] Update diagnostic message in 'Syntax/Parser/diags.swift'

beccadax committed:
Adjust to simulator compiler_rt change

The compiler_rt static library for iOS simulator is now packaged in a separate static archive from the one for iOS devices. Adjust linker invocation to match.

Fixes rdar://66192830 / rdar://66060312.

beccadax committed:
[NFC] Clean up unused simulator special cases

The previous commit makes various simulator-specific platform helpers dead code. Remove them.

xedin committed:
[AST] Mark repr invalid only if `@autoclosure` parameter doesn't point to function type

Instead on depending on repr to be a function, let's only check
whether type resolved for `@autoclosure` points to a function type
because it's allowed for `@autoclosure` parameters to be to
wrapped into parens or be represented by a typealias.

Resolves: rdar://problem/65704049

compnerd committed:
IRGen: always execute the coroutine lowering pass

Do not gate the coroutine extension points on the LLVM passes. Without
running this pass, the generated IR cannot be used by the LLVM tooling.
This allows generating the LLVM IR to debug issues in the LLVM backend.

I encountered this when trying to isolate a debug info generation bug
which seems to be caused by the SRoA pass in LLVM. By allowing to emit
the LLVM IR without the LLVM optimizations, it is possible to isolate
the LLVM pass operation via `opt` from LLVM.

slavapestov committed:
Sema: Remove GenericRequirementsCheckListener

tbkka committed:
Disable a hack that causes breakage on Python 3.8 (#33137)

Hacking sys.modules here was added unconditionally to fix an import problem on
Windows (with Python 2.7???).

This script works fine on Python 2.7 on macOS either with or without this hack.

This script breaks badly on Python 3.8 on macOS with this hack, so I've disabled
it here for all Python 3.

hborla committed:
[Function Builders] Teach diagnostics about function builder body
result types in order to properly diagnose requirement failures that
aren't anchored at an expression.

zoecarver committed:
[cxx-interop] [nfc] HasMemberWithDestructor => StructWithSubobjectDestructor in comment.

The type "HasMemberWithDestructor" doesn't exist. It was replaced with
StructWithSubobjectDestructor but the comment wasn't updated.

aschwaighofer committed:
Move the EagerSpecializer pass to the Transforms directory

It is no longer a module transform.

theblixguy committed:
[Sema] Diagnose wrapped property if its projected value property conflicts with lazy variable storage property (#33152)

natecook1000 committed:
XFAIL SILOptimizer/large_string_array.swift.gyb (#33157)

DougGregor committed:
Fix tests for slightly altered Fix-Its for throws/rethrows parsing.

DougGregor committed:
[Concurrency] Rename `-enable-experimental-async`.

Use the more general name `-enable-experimental-concurrency` instead,
since async is one part of the model.

DougGregor committed:
[Concurrency] Improve comments based on review feedback.

natecook1000 committed:
Fix swift_reflectionMirror_recursive* when called on pure ObjC classes. (#33143)

hamishknight committed:
Introduce LoweredSILRequest

Add a request that produces lowered SIL for a file
or module and use it in IRGenRequest if no
SILModule is provided.

theblixguy committed:
[Sema] Diagnose explicit access to a lazy variable's underlying storage (#33144)

gottesmm committed:
[sil] Add an implicit operator bool conversion to SILDebugLocation.

The reason to do this is that one can now do:

```
if (auto debugLoc = i.getDebugLocation()) {
  // do something with debug loc.
}
```

I also cleaned up the style a little bit by renaming the internal ivars to be
camelCase.

tbkka committed:
[Python3] Fix "undefined symbol 'unicode'" from python_lint (#33146)

* [Python3] Fix "undefined symbol 'unicode'" from python_lint

This is a little tricky.

Python 2 "unicode" was renamed to "str" in Python 3.

For Python 2 compatibility, we need to use "unicode" in a couple
of places, but that's not defined on Python 3, which causes
python_lint errors (even if the reference is never actually executed).

To workaround this, when running in Python 3, define "unicode"
as a synonym for "str".  This defines the symbol (avoiding the
"undefined symbol" error from python lint) while preserving
the correct functionality on both Python 2 and Python 3.

When we drop Python 2 support (which we should do as soon as
possible), we can drop this workaround and globally replace
"unicode" with "str" to get the right Python 3-only functionality.

tbkka committed:
[Python3] Fix test failures in update_checkout (#33141)

Use correct relative import for Python 2&3

Remove unused variable

Fix import ordering per python_lint

gottesmm committed:
[silgen] Mark a bunch of functions that SILGen open-codeds as "auto-generated" and thus not user generated code.

This ensures that we do not emit diagnostics onto these functions. I noticed we
were doing this with my work on opt-remarks when I was hitting diagnostics in
these functions in a set of subsequent commits.

The way that I found these is that I went through this file and looked for any
cases where we were never calling emitStmt on a user provided function body.

I also eliminated two places where we were setting a new DebugScope even though
preEmitFunction already does that for us.

DougGregor committed:
[Concurrency] Add Syntax support for 'async'.

Add 'async' to the syntax tree for types, function declarations, and
the "arrow" expression.

compnerd committed:
test: sort the environment for `remote-run`

With Python 3, there have been some sorting inconsistencies.  Always
sort the items to allow for the lit test to reliably match the output.

hborla committed:
[ConstraintLocator] Add a `isFirstElement` helper function

vedantk committed:
Add _getMetadataSection{,Count,Name} to API digester allow list (#33138)

These APIs are needed to build SwiftReflectionTest.swift when testing
Release builds.

This rolls back 014918c0, which hid these APIs in non-asserts builds
causing a failure:

  https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android-arm64/5708/consoleText

rdar://66103895

DougGregor committed:
[Concurrency] Finish and add tests for ban on `async` with `@objc`.

artemcm committed:
Change ExplicitModuleInfo to have String members instead of StringRef

gottesmm committed:
[opt-remark] Print out the type when we retain/release.

This can be useful if we can't recognize a retain/release and one needs to
reason about what is being retained/released.

compnerd committed:
test: normalize the path before opening

The tests mix path separators when passing paths to `gyb`.  Ensure that
we normalize the path string to allow the use of mixed path separators.

DougGregor committed:
[Concurrency] Minor fixes for parsing and syntax of 'async'

artemcm committed:
Use .str() when converting StringRef to std::string in ModuleInterfaceLoader

hyp committed:
Add missing `-f` to check which checks if a compiler-rt archive is present

DougGregor committed:
[Function builders] Fix typo in comment.

CodaFi committed:
Allow Deriving Type Witnesses Across Files

A recent change to witness matching in #32578 suddenly made the
following construction illegal:

// File1.swift
enum MyEnumInAnotherFile { /**/ }

// File2.swift
extension MyEnumInAnotherFile {
  static var allCases: [MyEnumInAnotherFile] { /**/ }
}

Because it was no longer possible to derive the type witness for
`AllCases`. This is because, when inference ran before synthesis, we
would use the value witness to pick out the type witness and thus had no
need for synthesis. Now that we run synthesis first, we ought to just
allow deriving type witnesses across files, but still ban deriving value
witnesses. In general, if you can utter a type in a different file to
extend it, you should be able to see the requirements necessary to
derive a default type witness.

rdar://66279278, rdar://66279375, rdar://66279384, rdar://66279415, rdar://66279503

jckarter committed:
IRGen: Give ObjC metadata symbols internal linkage.

This preserves the symbol names in (unstripped) binaries, so we can use them for link ordering and
memory usage analysis.

beccadax committed:
Mark main() as used

This prevents dead code stripping from removing the main() function used in widgets. This special case was added to support the integrated REPL, which was removed after Swift 5.3, and so it is no longer necessary.

Fixes rdar://65862827.

artemcm committed:
Special-case catalyst when computing a path for stdlib_dir in lit config
target_sdk_name is `macosx` when targeting `maccatalyst`, but the stdlib swift modules are placed under `swift/maccatalyst` instead of `swift/macosx`.
Using `run_os` is not correct in the general sense though, hence the requirement for a special case instead of just replacing `target_sdk_name` with it.
For example, it would be invalid to use run_os instead of target_sdk_name on Windows, because the sdk_name is going to be windows while the run_os variable may be windows-msvc.

DougGregor committed:
[Concurrency] Stub out an experimental concurrency support library.

The experimental concurrency model will require a supporting runtime
and possibly end-user-visible library constructs. Introduce a stub of
such a library, enabled by a new `build-script` option
`--enable-experimental-concurrency`, so we have a place to put this
work.

theblixguy committed:
[CSDiagnostics] Insert '@escaping' fix-it when type has '@autoclosure' during non-escaping function type diagnostic (#33191)

augusto2112 committed:
Conditionalize 'num_extra_inhabitants' in 64 bit platforms' tests (#33136)

akyrtzi committed:
[SourceKit] Fix issue where `CompletionCheckDependencyInterval` is set to 0 when the global config request is sent

Fixes rdar://66309544

DougGregor committed:
[Concurrency] Update expected options for --enable-experimental-concurrency

rintaro committed:
[CodeCompletion] Fix issue in dependency checking for the editing file

'SM.getCodeCompletionBufferID()' returns the buffer ID of the previous
code completion. 'CI.getCodeCompletionFile()->getBufferID()' always
returns the original source buffer ID of the current file in the main
module. The latter is needed for excluding buffer ID for dependency checking.

rdar://problem/66301353

eeckstein committed:
StringOptimization: bug fixes

Fix some bugs in the optimization and in the test files.

rdar://problem/66283894

eeckstein committed:
StringOptimization: fix a test for linux and windows.

DougGregor committed:
Appease pointless Python lint error

rintaro committed:
[Tests] Disable failing test cases while investigating

rdar://problem/66330768

aschwaighofer committed:
Type substitution: When substituting SILFunctionTypes we substitute unbound Objective-C generic for bound ones (take 2)

IRGen does so. So don't assert on the case when there is a nested
Objective-C type.

(follow-up to #32027 rdar://63509292 which fixed the non-nested case)

rdar://66302139

dan-zheng committed:
[AutoDiff] Fix class differentiation. (#33151)

Fix pullback generation for `unchecked_ref_cast` and `upcast` instructions
for class types with an address tangent value category.

Upstream end-to-end class differentiation tests.

tbkka committed:
Use `assert(mock.called)` (#33194)

The `mock.assert_called()` method was added in Python 3.6,
so fails on some of our older CI machines that use Python 3.5.
`assert(mock.called)` is precisely equivalent and works with
older versions of the Python unittest.mock library.

nkcsgexi committed:
Front-end: pass down main executable path to the sub-ASTContext for building module interfaces. NFC

tbkka committed:
[Python3] Use correct Python version for LLDB-related tests (#33181)

* Use the correct Python executable to match LLDB

The linux-fatal-backtrace script needs to run with an LLDB
module loaded into Python.  This in turn requires that the
test be run with the exact same Python version as was used
by LLDB (not just the same Python module directory).  This
can get confused on systems with multiple versions of Python
installed.

This replaces `lldb-python-path` (the Python module directory path)
with `lldb-python` (which is the correct Python version run with
the LLDB path).  This should ensure that this test is always
run with the same Python version and module that LLDB used.

* Use consistent braces for the lldb-python substitution

* Use os.path utilities to dissect paths

* Allow `python3` as a path identifier

Previous code required a decimal point before it would
recognize a path component as a Python version identifier,
so it would accept `python2.7` but not `python3`.

rintaro committed:
[Tests] SILOptimizer/opt-remark-generator.swift requires optimzed stdlib

rdar://problem/66332541

rintaro committed:
[Tests] Disable failing test while investigating

test/IRGen/objc_protocol_extended_method_types.swift

rdar://problem/66353882

rintaro committed:
[Tests] Disable Runtime/linux-fatal-backtrace.swift

rdar://problem/66283479

Commit:
edit

Commit:
Update ParseType.cpp
